File: typescript/src/Secrets/SecretsManager.ts
/**
 * Zirve Secrets Manager — Infisical Integration.
 *
 * Retrieves parameters and secrets from Infisical with an in-memory cache layer.
 */
export class SecretsManager {
  private config: Record<string, any>;
  private baseUrl: string;
  private token: string;
  private projectId: string;
  
  // In-memory local cache for quick secret access
  private cache: Record<string, { value: string; expiresAt: number }> = {};
  private readonly CACHE_TTL_MS = 300000; // 5 minutes

  constructor(config: Record<string, any>) {
    this.config = config;
    this.baseUrl = (config.url || 'http://infisical.zirve-infra.svc.cluster.local').replace(/\/$/, '');
    this.token = config.token || '';
    this.projectId = config.project_id || '';
  }

  /**
   * Get a secret by its key name from Infisical.
   */
  public async get(secretName: string, environment: string = 'dev', path: string = '/'): Promise<string | null> {
    const cacheKey = `${environment}:${path}:${secretName}`;

    // Check cache
    if (this.cache[cacheKey] && this.cache[cacheKey].expiresAt > Date.now()) {
      return this.cache[cacheKey].value;
    }

    if (!this.token || !this.projectId) {
      throw new Error('Infisical token and project ID must be configured.');
    }

    const url = new URL(`${this.baseUrl}/api/v3/secrets/raw/${secretName}`);
    url.searchParams.append('workspaceId', this.projectId);
    url.searchParams.append('environment', environment);
    url.searchParams.append('secretPath', path);

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${this.token}`,
      },
    });

    if (response.status === 404) {
      return null;
    }

    if (!response.ok) {
      throw new Error(`Infisical fetch failed: ${response.statusText}`);
    }

    const data = await response.json();
    const value = data.secret.secretValue;

    // Cache the secret
    this.cache[cacheKey] = {
      value,
      expiresAt: Date.now() + this.CACHE_TTL_MS,
    };

    return value;
  }

  /**
   * List all secrets in a specific environment and path.
   */
  public async list(environment: string = 'dev', path: string = '/'): Promise<Record<string, string>> {
    const url = new URL(`${this.baseUrl}/api/v3/secrets/raw`);
    url.searchParams.append('workspaceId', this.projectId);
    url.searchParams.append('environment', environment);
    url.searchParams.append('secretPath', path);

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Infisical list failed: ${response.statusText}`);
    }

    const data = await response.json();
    const result: Record<string, string> = {};

    for (const secret of data.secrets) {
      result[secret.secretKey] = secret.secretValue;
    }

    return result;
  }

  /**
   * Basic health check.
   */
  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/health`, {
        method: 'GET',
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}

File: dotnet/Zirve.Sdk/Secrets/SecretsManager.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;
using System.Collections.Generic;
using Zirve.Sdk.Config;
using System.Collections.Concurrent;

namespace Zirve.Sdk.Secrets;

public class SecretsManager
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private readonly string _token;
    private readonly string _projectId;

    private readonly ConcurrentDictionary<string, (string Value, DateTime ExpiresAt)> _cache = new();
    private static readonly TimeSpan CacheTtl = TimeSpan.FromMinutes(5);

    public SecretsManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("secrets");
        _baseUrl = (cfg.GetValueOrDefault("url", "http://infisical.zirve-infra.svc.cluster.local") ?? "").TrimEnd('/');
        _token = cfg.GetValueOrDefault("token", "") ?? "";
        _projectId = cfg.GetValueOrDefault("project_id", "") ?? "";
    }

    public async Task<string?> GetAsync(string secretName, string environment = "dev", string path = "/")
    {
        var cacheKey = $"{environment}:{path}:{secretName}";

        if (_cache.TryGetValue(cacheKey, out var cached) && cached.ExpiresAt > DateTime.UtcNow)
        {
            return cached.Value;
        }

        if (string.IsNullOrEmpty(_token) || string.IsNullOrEmpty(_projectId))
        {
            throw new InvalidOperationException("Infisical token and project ID must be configured.");
        }

        var uri = $"{_baseUrl}/api/v3/secrets/raw/{secretName}?workspaceId={_projectId}&environment={environment}&secretPath={Uri.EscapeDataString(path)}";
        var request = new HttpRequestMessage(HttpMethod.Get, uri);
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);

        var response = await _httpClient.SendAsync(request);

        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            return null;
        }

        response.EnsureSuccessStatusCode();

        using var jsonDoc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        var value = jsonDoc.RootElement.GetProperty("secret").GetProperty("secretValue").GetString();

        if (value != null)
        {
            _cache[cacheKey] = (value, DateTime.UtcNow.Add(CacheTtl));
        }

        return value;
    }

    public async Task<Dictionary<string, string>> ListAsync(string environment = "dev", string path = "/")
    {
        var uri = $"{_baseUrl}/api/v3/secrets/raw?workspaceId={_projectId}&environment={environment}&secretPath={Uri.EscapeDataString(path)}";
        var request = new HttpRequestMessage(HttpMethod.Get, uri);
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        using var jsonDoc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        var secrets = jsonDoc.RootElement.GetProperty("secrets").EnumerateArray();

        var result = new Dictionary<string, string>();
        foreach (var secret in secrets)
        {
            var key = secret.GetProperty("secretKey").GetString();
            var val = secret.GetProperty("secretValue").GetString();
            if (key != null && val != null)
            {
                result[key] = val;
            }
        }

        return result;
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_baseUrl}/api/v1/health");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}
